tools/autograd/load_derivatives.py

```diff
@@ -315,6 +315,17 @@ def saved_variables(formula, args):
             'suffix': '_dim',
             'type': 'int64_t',
         }),
+        # replace self.is_sparse() with self_is_sparse
+        (r'{}.is_sparse\(\)', {
+            'suffix': '_is_sparse',
+            'type': 'bool',
+        }),
+        # replace self.strides() with self_strides
+        (r'{}.strides\(\)', {
+            'suffix': '_strides',
+            'type': 'IntArrayRef',
+            'expr': lambda name: f'({name}.is_sparse() ? IntArrayRef() : {name}.strides())',
```

Comment:
I had to do this because if the `{name}` tensor is sparse, calling `.strides()` on it produces an error: `RuntimeError: sparse tensors do not have strides`. While it's true that the `mm_mat1_backward` call will produce an error if `mat1` is sparse anyway, the error message is more specific, which helps with debugging: `RuntimeError: calculating the gradient of a sparse Tensor argument to mm is not supported.`

However, this solution is unsafe, because some derivative formulas probably should throw the `RuntimeError: sparse tensors do not have strides` error when `strides()` is called. With this solution, this case would incorrectly return an empty strides array.

Perhaps a better solution could be to remove the `mat1_is_sparse` check from `mm_mat1_backward`, and then in `derivatives.yaml`, we can have an expression that tells `load_derivatives.py` to replace `mat1.strides()` with some code that throws a specified error message if `mat1` is sparse. Maybe it would look something like this:

```
- name: addmm(Tensor self, Tensor mat1, Tensor mat2, *, Scalar beta=1, Scalar alpha=1) -> Tensor
  self: maybe_multiply(grad, beta)
  mat1: mm_mat1_backward(grad, mat2, mat1.sizes(), strides_or_error_if_sparse(mat1, "calculating the gradient of a sparse Tensor argument to mm is not supported") , alpha)
  mat2: mm_mat2_backward(grad, mat1, mat2.sizes(), mat2.strides(), alpha)
```

`strides_or_error_if_sparse` could be a real function in `torch/csrc/autograd/FunctionsManual.cpp`. When `tools/autograd/load_derivatives.py` sees this expression, it can still replace it with `mat1_strides`, but the generated code in `torch/csrc/autograd/generated/VariableType_0.cpp` would be:

```
grad_fn->mat1_strides = strides_or_error_if_sparse(mat1, "calculating the gradient of a sparse Tensor argument to mm is not supported").vec();
```

If I understand correctly, this is doable and wouldn't require too much extra work, so I will try it out.

Reply:
I was able to come up with a solution. Probably could be improved though.